EOS-27895 : Create granular images

gauravchaudhari02 opened the pull request:
# Problem Statement
- Previously, cortx-all image was used which has all cortx components. now we need to create separate images based on components.
# Design
-  Create multiple images based on components using new folder structure.

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
   cortx-data : http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/nikhildev/job/cortx-granular-docker-image/44/
   cortx-control : http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/nikhildev/job/cortx-granular-docker-image/42/
   cortx-ha : http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/nikhildev/job/cortx-granular-docker-image/40/
   cortx-all : http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/nikhildev/job/cortx-granular-docker-image/15/
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

shailesh-vaidya commented on docker/cortx-deploy/cortx-all/third-party-rpms.txt:
```diff
@@ -2,6 +2,7 @@ crontabs  #cortx-utils
 haproxy22 #cortx-s3srever
 java-1.8.0-openjdk-headless
 jq #cortx-prvsnr
+kafka  #cortx-utils
```
Kafka is removed as part of https://github.com/Seagate/cortx-re/pull/803

gauravchaudhari02 replied:
Removed kafka package

shailesh-vaidya commented on docker/cortx-deploy/build.sh:
```diff
@@ -63,12 +65,24 @@ echo "Building $SERVICE image from $BUILD_URL"
-echo cortx-csm_agent:"$(awk -F['_'] '/cortx-csm_agent-'$VERSION'/ { print $3 }' RELEASE.INFO | cut -d. -f1)",
+        sed -i '/KERNEL/d' RELEASE.INFO
+        if [ "$SERVICE" == "cortx-data" ]; then 
+                COMPONENTS="cortx-hare cortx-motr cortx-py-utils cortx-provisioner"
```
Should we get a component list from service specific  `cortx-componenet-rpms.txt` to keep a single point of contact for CORTX component update.

gauravchaudhari02 replied:
Updated the logic as per ask and tested - http://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/nikhildev/job/cortx-granular-docker-image/34/

shailesh-vaidya commented on docker/cortx-deploy/cortx-control/Dockerfile:
```diff
@@ -51,22 +51,21 @@ RUN yum-config-manager --add-repo ${BUILD_URL}/cortx_iso/ \
 RUN yum install --nogpgcheck -y python3-pip \
     && pip3 install yq \
-    && pip3 install  --no-cache-dir --trusted-host $(echo $BUILD_URL | awk -F '/' '{print $3}') -i $BUILD_URL/python_deps -r /opt/seagate/cortx/python_requirements.txt -r /opt/seagate/cortx/python_requirements.ext.txt \
+    && pip3 install  --no-cache-dir --trusted-host "$(echo $BUILD_URL | awk -F '/' '{print $3}')" -i $BUILD_URL/python_deps -r /opt/seagate/cortx/python_requirements.txt \
```
Do we not need `/opt/seagate/cortx/python_requirements.ext.txt` . Just in case python package has not declared dependency in there package manegemnet.

gauravchaudhari02 replied:
@shailesh-vaidya, Do you mean we should keep `/opt/seagate/cortx/python_requirements.ext.txt` incase component need to declare some deendencies which are not covered in `/opt/seagate/cortx/python_requirements.txt`?

shailesh-vaidya replied:
Yes. So if dependencies are not installed automatically then install it form `/opt/seagate/cortx/python_requirements.ext.txt`

gauravchaudhari02 replied:
Calling `/opt/seagate/cortx/python_requirements.ext.txt` file from `/opt/seagate/cortx/python_requirements.txt` with `-r` flag.

shailesh-vaidya reviewed (changes requested):
- Do we need to change at - https://github.com/Seagate/cortx-re/blob/main/jenkins/automation/kubernetes/cortx-all-image.groovy#L90 
- docker/cortx-deploy/common-python-req.txt - This seems to be an empty file. 
- Component teams have provided respective third-party packages so we review that list and keep tuning it. e.g. We can remove `cortx-s3srever` from all images. 
- Can we have a single file for third-party packages with component names tagged. So we can select a third-party package to install based on its tag. This will also reduce complexity in adding third-party packages.

gauravchaudhari02 commented:
> * Do we need to change at - https://github.com/Seagate/cortx-re/blob/main/jenkins/automation/kubernetes/cortx-all-image.groovy#L90 

Yes, We have separate PR https://github.com/Seagate/cortx-re/pull/849 to address this. 

> * docker/cortx-deploy/common-python-req.txt - This seems to be an empty file.

Removed it.

> * Component teams have provided respective third-party packages so we review that list and keep tuning it. e.g. We can remove `cortx-s3srever` from all images.

Removed `cortx-s3server` 

> * Can we have a single file for third-party packages with component names tagged. So we can select a third-party package to install based on its tag. This will also reduce complexity in adding third-party packages.

Will take it as improvement.

shailesh-vaidya commented:
> > * docker/cortx-deploy/common-python-req.txt - This seems to be an empty file.
> 
> Removed it.
> 
Seems it's still present - https://github.com/gauravchaudhari02/cortx-re/blob/create-granular-images/docker/cortx-deploy/common-python-req.txt
